feat(formula): align to template formula

noelmcloughlin opened the pull request:
<!--
Please fill in this PR template to make it easier to review and merge.

It has been designed so that a lot of it can be completed *after* submitting it,
e.g. filling in the checklists.

Notes:
1. Please keep the PR as small as is practicable; the larger a PR gets, the harder it becomes to review and the more time it requires to get it merged.
2. Similarly, please avoid PRs that cover more than one type; it should be a _bug fix_ *OR* a _new feature_ *OR* a _refactor_, etc.
3. Please direct questions to the [`#formulas` channel on Slack](https://saltstackcommunity.slack.com/messages/C7LG8SV54/), which is bridged to `#saltstack-formulas` on Freenode.
4. Feel free to suggest improvements to this template by reporting an issue or submitting a PR.  The source of this template is:
  - https://github.com/saltstack-formulas/.github/blob/master/.github/pull_request_template.md
-->

### PR progress checklist (to be filled in by reviewers)
<!-- Please leave this checklist for reviewers to tick as they work through the PR. -->

- [x] Changes to documentation are appropriate (or tick if not required)
- [x] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

### What type of PR is this?
<!-- Please tick each box that is relevant (after creating the PR). -->

#### Primary type
<!-- There really should be only *one* of these types ticked for each PR. -->

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [ ] `[ci]`       Changes to the continuous integration configuration
- [x] `[feat]`     A new feature
- [x] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [x] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [x] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type
<!-- Most PRs should include all of the following types as well. -->

- [x] `[docs]`     Documentation changes
- [x] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?
<!-- If so, change the following to a `Yes` and explain what the breaking changes are. -->
<!-- If there are multiple breaking changes, list them all. -->

YES

### Related issues and/or pull requests
<!-- Please link any related issues/PRs here, especially any issues that are closed by this PR. -->



### Describe the changes you're proposing
<!-- A clear and concise description of what you have implemented. -->
<!-- Consider explaining each commit if they cover different aspects of the proposed changes. -->

Using IntelliJ-formula as a template, align to template-formula.

### Pillar / config required to test the proposed changes
<!-- Provide links to the SLS files and/or relevant configs (be sure to remove sensitive info). -->



### Debug log showing how the proposed changes work
<!-- Include a debug log showing how these changes work, e.g. using `salt-minion -l debug`. -->
<!-- Alternatively, linking to Kitchen debug logs is useful, e.g. via. Travis CI. -->
<!-- Most useful is providing a passing InSpec test, which can be used to verify any proposed changes. -->



### Documentation checklist
<!-- Please tick each box that is relevant (after creating the PR). -->

- [x] Updated the `README` (e.g. `Available states`).
- [x] Updated `pillar.example`.

### Testing checklist
<!-- Please tick each box that is relevant (after creating the PR). -->

- [x] Included in Kitchen (i.e. under `state_top`).
- [x] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [x] Updated the relevant test pillar.

### Additional context
<!-- Add any other context about the proposed changes here. -->

noelmcloughlin commented:
Trace on MacOS
```
  ID: pycharm-macos-app-clean-files
    Function: file.absent
        Name: /tmp/jetbrains/pycharm
      Result: True
     Comment: File /tmp/jetbrains/pycharm is not present
     Started: 14:49:26.990998
    Duration: 0.512 ms
     Changes:   
----------
          ID: pycharm-macos-app-clean-files
    Function: file.absent
        Name: /Applications/Pycharm.app
      Result: True
     Comment: File /Applications/Pycharm.app is not present
     Started: 14:49:26.991644
    Duration: 0.342 ms
     Changes:   
----------
          ID: pycharm-config-clean-file-absent
    Function: file.absent
        Name: /tmp/dummy_list_item
      Result: True
     Comment: File /tmp/dummy_list_item is not present
     Started: 14:49:26.992197
    Duration: 0.352 ms
     Changes:   
----------
          ID: pycharm-config-clean-file-absent
    Function: file.absent
        Name: /etc/defaults/pycharm.sh
      Result: True
     Comment: File /etc/defaults/pycharm.sh is not present
     Started: 14:49:26.992799
    Duration: 0.942 ms
     Changes:   
----------
          ID: pycharm-config-clean-file-absent
    Function: file.absent
        Name: /Users/nmcloughlin/Desktop/Pycharm CE
      Result: True
     Comment: Removed file /Users/nmcloughlin/Desktop/Pycharm CE
     Started: 14:49:26.993984
    Duration: 12.068 ms
     Changes:   
              ----------
              removed:
                  /Users/nmcloughlin/Desktop/Pycharm CE
----------
          ID: pycharm-macos-app-install-curl
    Function: file.directory
        Name: /tmp/jetbrains/pycharm
      Result: True
     Comment: Directory /tmp/jetbrains/pycharm updated
     Started: 14:49:27.006220
    Duration: 2.269 ms
     Changes:   
              ----------
              /tmp/jetbrains/pycharm:
                  New Dir
----------
          ID: pycharm-macos-app-install-curl
    Function: pkg.installed
        Name: curl
      Result: True
     Comment: All specified packages are already installed
     Started: 14:49:31.626974
    Duration: 11248.389 ms
     Changes:   
----------
          ID: pycharm-macos-app-install-curl
    Function: cmd.run
        Name: curl -Lo /tmp/jetbrains/pycharm/pycharm-2020.1 https://download.jetbrains.com/python/pycharm-community-2020.1.dmg
      Result: True
     Comment: Command "curl -Lo /tmp/jetbrains/pycharm/pycharm-2020.1 https://download.jetbrains.com/python/pycharm-community-2020.1.dmg" run
     Started: 14:49:42.879344
    Duration: 666600.319 ms
     Changes:   
              ----------
              pid:
                  5338
              retcode:
                  0
              stderr:
                    % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                                   Dload  Upload   Total   Spent    Left  Speed
                  
                    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
                    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
                  100   138  100   138    0     0    309      0 --:--:-- --:--:-- --:--:--   308
                  
                    0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
                    0  360M    0  117k    0     0  56392      0  1:51:37  0:00:02  1:51:35  117k
                    0  360M    0  660k    0     0   193k      0  0:31:46  0:00:03  0:31:43  289k
     
       << CUT >>

                   99  360M   99  359M    0     0   552k      0  0:11:07  0:11:05  0:00:02  672k
                   99  360M   99  359M    0     0   552k      0  0:11:07  0:11:06  0:00:01  635k
                  100  360M  100  360M    0     0   553k      0  0:11:06  0:11:06 --:--:--  741k
              stdout:
----------
          ID: pycharm-macos-app-install-checksum
    Function: module.run
        Name: file.check_hash
      Result: True
     Comment: onlyif condition is false
     Started: 15:00:49.485185
    Duration: 909.815 ms
     Changes:   
----------
          ID: pycharm-macos-app-install-checksum
    Function: file.absent
        Name: /tmp/jetbrains/pycharm/pycharm-2020.1
      Result: True
     Comment: State was not run because onfail req did not change
     Started: 15:00:50.396295
    Duration: 0.007 ms
     Changes:   
----------
          ID: pycharm-macos-app-install-macpackage
    Function: macpackage.installed
        Name: /tmp/jetbrains/pycharm/pycharm-2020.1
      Result: True
     Comment: PyCharm CE.app installed
     Started: 15:00:50.396641
    Duration: 25264.116 ms
     Changes:   
              ----------
              installed:
                  - PyCharm CE.app
----------
          ID: pycharm-macos-app-install-macpackage
    Function: file.managed
        Name: /tmp/mac_shortcut.sh
      Result: True
     Comment: File /tmp/mac_shortcut.sh updated
     Started: 15:01:15.661058
    Duration: 42.897 ms
     Changes:   
              ----------
              diff:
                  --- 
                  +++ 
                  @@ -1,5 +1,5 @@
                   #!/usr/bin/env bash
                   
                  -Source="/Applications/Intellij IDEA CE.app"
                  +Source="/Applications/Pycharm CE.app"
                   Destination="/Users/nmcloughlin/Desktop"
                   /usr/bin/osascript -e "tell application \"Finder\" to make alias file to POSIX file \"$Source\" at POSIX file \"$Destination\""
----------
          ID: pycharm-macos-app-install-macpackage
    Function: cmd.run
        Name: /tmp/mac_shortcut.sh
      Result: True
     Comment: Command "/tmp/mac_shortcut.sh" run
     Started: 15:01:15.704580
    Duration: 1397.742 ms
     Changes:   
              ----------
              pid:
                  7577
              retcode:
                  0
              stderr:
              stdout:
                  alias file PyCharm CE of folder Desktop of folder nmcloughlin of folder Users of startup disk
----------
          ID: pycharm-config-file-file-managed-environ_file
    Function: file.managed
        Name: /etc/defaults/pycharm.sh
      Result: True
     Comment: File /etc/defaults/pycharm.sh updated
     Started: 15:01:17.102864
    Duration: 1660.913 ms
     Changes:   
              ----------
              diff:
                  New file
              group:
                  nmcloughlin
              mode:
                  0644
              user:
                  nmcloughlin

Summary for local
-------------
Succeeded: 14 (changed=7)
Failed:     0
-------------
Total states run:     14
```

myii commented:
@noelmcloughlin I've finished checking this with the `ssf-formula`:

* https://github.com/myii/ssf-formula/pull/180

I've run it in my own fork with the standardised changes, including updating to the latest images -- all is working well:

* https://travis-ci.com/github/myii/jetbrains-pycharm-formula/builds/164853358

Here's the actual commit to be added here:

* https://github.com/myii/jetbrains-pycharm-formula/commit/a3bac46edc0678d4436b425e3929d5c04aea72f6
  - I've got some feedback to share about this, I still need to decide whether to add it here or as comments on that commit.

I wasn't able to add it here because of permissions (deselected maintainer edits?).

There's also the merge conflicts to be resolved here due to #7; it would be helpful if PRs were commenced from the latest commit in `master` (or at least rebased afterwards) to avoid conflicts and ease reviewing/amendments.

myii commented on .travis.yml:
```diff
@@ -0,0 +1,119 @@
+        - gem install rubocop
+        - rubocop -d
+        # Run `shellcheck` (already pre-installed in Travis)
+        - shellcheck --version
```
Nothing being done with `shellcheck` here, will be resolved when commit is added here:

https://github.com/myii/jetbrains-pycharm-formula/commit/a3bac46edc0678d4436b425e3929d5c04aea72f6#diff-354f30a63fb0907d4ad57269548329e3R51-R52

noelmcloughlin replied:
Hi, @myii Thanks for the effort in reviewing and providing the feedback.

myii commented on .yamllint:
```diff
@@ -0,0 +1,39 @@
+  test/**/states/**/*.sls
+  .kitchen/
+  pycharm/osfamilymap.yaml
+  pycharm/files/default/shortcut.desktop.jinja
+  pillar.example
```
Unnecessary, will be resolved when commit is added here:

https://github.com/myii/jetbrains-pycharm-formula/commit/a3bac46edc0678d4436b425e3929d5c04aea72f6#diff-b97c55dd898003930e2858fadd3b5c9cL16-L17

myii commented on FORMULA:
```diff
@@ -1,8 +1,9 @@
-minimum_version: 2016.3
-summary: Formula for installing pycharm
-description: Formula for installing PyCharm IDE for Python developers
+minimum_version: 2016.11
```
We're no longer supporting < `2019.2`, will be resolved when commit is added here:

https://github.com/myii/jetbrains-pycharm-formula/commit/a3bac46edc0678d4436b425e3929d5c04aea72f6#diff-adc4995cd52208dde39433fb069ddd61R6

myii commented on Gemfile:
```diff
@@ -0,0 +1,11 @@
+gem 'kitchen-docker', '>= 2.9'
+gem 'kitchen-inspec', '>= 1.1'
+gem 'kitchen-salt', '>= 0.6.0'
+# Latest versions of `train` cause failure when running `kitchen verify`
+# Downgrading to `3.2.0` until this is fixed upstream
+# https://github.com/inspec/train/pull/544#issuecomment-566055052
+gem 'train', '3.2.0'
```
No longer necessary since fixed upstream and we're also adding the `Gemfile.lock` now, will be resolved when commit is added here:

https://github.com/myii/jetbrains-pycharm-formula/commit/a3bac46edc0678d4436b425e3929d5c04aea72f6#diff-8b7db4d5cc4b8f6dc8feb7030baa2478L8-L11

myii commented on docs/README.rst:
```diff
@@ -0,0 +1,163 @@
+.. _readme:
+
+pycharm-formula
+============
```
Headings markers should match heading text length, will be resolved when commit is added here:

https://github.com/myii/jetbrains-pycharm-formula/commit/a3bac46edc0678d4436b425e3929d5c04aea72f6#diff-8ab5636c6c18641e5c58358ef489e0f7R4

myii commented on docs/README.rst:
```diff
@@ -0,0 +1,163 @@
+
+Formula to install Pycharm on GNU/Linux and MacOS.
+
+.. contents:: **Table of Contents**
+
```
Using `depth` after your PR to the `template-formula`(!), will be resolved when commit is added here:

https://github.com/myii/jetbrains-pycharm-formula/commit/a3bac46edc0678d4436b425e3929d5c04aea72f6#diff-8ab5636c6c18641e5c58358ef489e0f7R20

myii commented on docs/TOFS_pattern.rst:
```diff
@@ -0,0 +1,518 @@
+       - require:
+         - pkg: Install NTP package
+
+If we want to cover the possibility of a special template for a minion identified by ``pycharm01`` then we could have a specific template in ``/srv/saltstack/salt/ntp/files/pycharm01/etc/ntp.conf.jinja``.
+
+.. code-block:: jinja
+
+   {#- /srv/saltstack/salt/ntp/files/pycharm01/etc/ntp.conf.jinja #}
+   {#- Managed by saltstack #}
+   {#- Edit pillars or override this template in saltstack if you need customization #}
+
+   {#- Some crazy configurations here for pycharm01 #}
```
We use the same `docs/TOFS_pattern.rst` in all formulas, will be resolved when commit is added here:

https://github.com/myii/jetbrains-pycharm-formula/commit/a3bac46edc0678d4436b425e3929d5c04aea72f6#diff-bf838b5bad3aeeb4d43d3e986622d832R263-R273

myii commented on kitchen.yml:
```diff
@@ -0,0 +1,209 @@
+            '*':
+              - pycharm
+      pillars_from_files:
+        pycharm.sls: test/salt/default/pillar.sls
```
The general standard we decided upon for test pillars is:

* `test/salt/pillar/<suite-name-or-similar>.sls`
  - We've not been too picky about the name of the actual file, as long as it's clear
  - Personally, I believe it should always contain the InSpec suite name at the very least, if that's not already sufficient

You can see how this is being used across the org by searching for `pillars_from_files` in the `formulas.yaml` of the `ssf-formula`:

* https://github.com/myii/ssf-formula/blob/master/ssf/formulas.yaml

In this PR, that would also affect the line further below.

myii commented on kitchen.yml:
```diff
@@ -0,0 +1,209 @@
+    verifier:
+      inspec_tests:
+        - path: test/integration/default
+  - name: arch
+    provisioner:
```
Adding inclusions and exclusions (particularly inclusions) when there are multiple suites is important, otherwise it can lead to massive lists when running `kitchen list` -- and most of these won't even work due to the nature of the suites; will be resolved when commit is added here:

https://github.com/myii/jetbrains-pycharm-formula/commit/a3bac46edc0678d4436b425e3929d5c04aea72f6#diff-fdbca9ed07a634829a3548f874dc925aR146-R168

myii commented on test/integration/arch/README.md:
```diff
@@ -0,0 +1,50 @@
+# InSpec Profile: `default`
```
The actual suite name needs to be used throughout this file, will be resolved when commit is added here:

https://github.com/myii/jetbrains-pycharm-formula/commit/a3bac46edc0678d4436b425e3929d5c04aea72f6#diff-19a94498bc41afb14c871deec9e4b473L1-R43

myii commented on test/integration/arch/inspec.yml:
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+# vim: ft=yaml
+---
+name: default
+title: pycharm formula
+maintainer: SaltStack Formulas
+license: Apache-2.0
+summary: Verify that the pycharm formula is setup and configured correctly
```
The name and summary should be appropriate for each suite, will be resolved when commit is added here:

https://github.com/myii/jetbrains-pycharm-formula/commit/a3bac46edc0678d4436b425e3929d5c04aea72f6#diff-e12e128804f69778e0d42bd5d468d7a8L4-R9

noelmcloughlin commented:
Hi @myii 
I've gone to considerable effort to align all my new formulas (new PRs) and updated formulas (existing PRs) with all your comments.  The only one not actioned is changing `test/salt/<suite-name-or-similar>/pillar.sls` to `test/salt/pillar/<suite-name-or-similar>.sls` because I missed that comment.   That is something I can look at in future I guess. I fixed the conflicts.

This work should make reviews easier because the comments here are actioned in all my open PRs.

myii commented:
> I've gone to considerable effort to align all my new formulas (new PRs) and updated formulas (existing PRs) with all your comments. The only one not actioned is changing `test/salt/<suite-name-or-similar>/pillar.sls` to `test/salt/pillar/<suite-name-or-similar>.sls` because I missed that comment. That is something I can look at in future I guess. I fixed the conflicts.
> 
> This work should make reviews easier because the comments here are actioned in all my open PRs.

Thanks for your efforts, @noelmcloughlin.  I'll keep working my way through the PRs.

In the meantime, I've updated the commit to be added here:

* https://github.com/myii/jetbrains-pycharm-formula/commit/d2c58246ec5b07e0dd0b8038d8882854162ce00e
* https://travis-ci.com/github/myii/jetbrains-pycharm-formula/builds/165078462

I still couldn't add it here because of permissions.  So please could you add that commit here yourself, or make sure `Allow edits by maintainers` is selected and then let me know so that I can push it here.

noelmcloughlin commented:
Can you try now? I updated access  Allow edits by maintainers seems to be
ON now.

On Mon, May 11, 2020 at 3:18 AM Imran Iqbal <notifications@github.com>
wrote:

> I've gone to considerable effort to align all my new formulas (new PRs)
> and updated formulas (existing PRs) with all your comments. The only one
> not actioned is changing test/salt/<suite-name-or-similar>/pillar.sls to
> test/salt/pillar/<suite-name-or-similar>.sls because I missed that
> comment. That is something I can look at in future I guess. I fixed the
> conflicts.
>
> This work should make reviews easier because the comments here are
> actioned in all my open PRs.
>
> Thanks for your efforts, @noelmcloughlin
> <https://github.com/noelmcloughlin>. I'll keep working my way through the
> PRs.
>
> In the meantime, I've updated the commit to be added here:
>
>    - myii@d2c5824
>    <https://github.com/myii/jetbrains-pycharm-formula/commit/d2c58246ec5b07e0dd0b8038d8882854162ce00e>
>    -
>    https://travis-ci.com/github/myii/jetbrains-pycharm-formula/builds/165078462
>
> I still couldn't add it here because of permissions. So please could you
> add that commit here yourself, or make sure Allow edits by maintainers is
> selected and then let me know so that I can push it here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack-formulas/jetbrains-pycharm-formula/pull/8#issuecomment-626435555>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADFUUQXYLH663KZJ73TKNRDRQ5OAHANCNFSM4MNK3K7A>
> .
>

myii commented:
@noelmcloughlin Still no cigar:

```bash
ERROR: Permission to noelmcloughlin/jetbrains-pycharm-formula.git denied to myii.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Not sure why it's being problematic, I've been using it fine elsewhere, as you know.  If you add my fork locally, it's pretty simple to cherry-pick my commit and push it here.

myii commented:
Thanks, @noelmcloughlin -- we finally got there.

Note, that `centos-8` job `Ran for 10 min 7 sec ` -- may need to keep an eye on that.  I prefer to set `retry: 1` for our Travis jobs (in the test pillar), to prevent them holding up other instances across the org, since we've only got 5 concurrent in total.

saltstack-formulas-travis commented:
:tada: This PR is included in version 1.0.0 :tada:

The release is available on [GitHub release](https://github.com/saltstack-formulas/jetbrains-pycharm-formula/releases/tag/v1.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
